vowpalwabbit/cats_tree.cc

```diff
@@ -0,0 +1,385 @@
+    {
+      const node_cost& n_c = set_d[i];
+      const tree_node& v = nodes[n_c.node_id];
+      const float cost_v = n_c.cost;
+      const tree_node& v_parent = nodes[v.parent_id];
+      float cost_parent = cost_v;
+      if (v_parent.right_only || v_parent.left_only)
+        continue;
+      const tree_node& w = _binary_tree.get_sibling(v);  // w is sibling of v
+      float cost_w = return_cost(w);
+      if (cost_v != cost_w)
+      {
+        VW_DBG(ec) << "otree_c: learn() cost_w = " << cost_w << ", cost_v != cost_w" << std::endl;
+        float local_action = RIGHT;
+        if (((cost_v < cost_w) ? v : w).id == v_parent.left_id)
+          local_action = LEFT;
+
+        ec.l.simple.label = local_action;
+        ec.l.simple.initial = 0.f;
+        ec.weight = abs(cost_v - cost_w);
+
+        bool filter = false;
+        const float weight_th = 0.00001f;
+        if (ec.weight < weight_th)
+        {
+          // generate a new seed
+          uint64_t new_random_seed = uniform_hash(&app_seed, sizeof(app_seed), app_seed);
+          // pick a uniform random number between 0.0 - .001f
+          float random_draw = exploration::uniform_random_merand48(new_random_seed)*weight_th;
+          if (random_draw < ec.weight) {
+            ec.weight = weight_th;
+          }
+          else {
+            filter = true;
+          }
+        }
+        if (!filter)
+        {
+          VW_DBG(ec) << "otree_c: learn() #### binary learning the node " << v.parent_id << std::endl;
+          base.learn(ec, v.parent_id);
+          _binary_tree.nodes[v.parent_id].learn_count++;
+          base.predict(ec, v.parent_id);
+          VW_DBG(ec) << "otree_c: learn() after binary predict:" << scalar_pred_to_string(ec)
+            << ", local_action = " << (local_action) << std::endl;
+          float trained_action = (ec.pred.scalar < 0) ? LEFT : RIGHT;
+          if (trained_action == local_action)
+          {
+            cost_parent =
+              (std::min)(cost_v, cost_w) * fabs(ec.pred.scalar) + (std::max)(cost_v, cost_w) * (1 - fabs(ec.pred.scalar));
+            VW_DBG(ec) << "otree_c: learn() ec.pred.scalar == local_action" << std::endl;
+          }
+          else
+          {
+            cost_parent =
+              (std::max)(cost_v, cost_w) * fabs(ec.pred.scalar) + (std::min)(cost_v, cost_w) * (1 - fabs(ec.pred.scalar));
+            VW_DBG(ec) << "otree_c: learn() ec.pred.scalar != local_action" << std::endl;
+          }
+        }
+
+      }
+      if (i == 0)
+        a_parent_cost = cost_parent;
+      else
+        b_parent_cost = cost_parent;
+    }
+    _a = {nodes[_a.node_id].parent_id, a_parent_cost};
+    _b = {nodes[_b.node_id].parent_id, b_parent_cost};
+  }
+
+  ec.l = saved_label;
+  ec.weight = saved_weight;
+  ec.pred = saved_pred;
+}
+
+void cats_tree::set_trace_message(std::ostream* vw_ostream) { _trace_stream = vw_ostream; }
+
+cats_tree::~cats_tree()
+{
+  if(_trace_stream != nullptr)
```

Comment:
Not sure if printing in the destructor is great, but I don't feel strongly here

Reply:
Yeah.  I agree.  I don't see a good place for this in the current workflow.